### ISSUE TITLE
no_std support

### ISSUE THREAD

**Issue by rnbguy** [Canop/lazy-regex#29]
Hey. Thanks for this crate :pray: 

It'd be useful if we had a `no_std` feature for `lazy-regex`, as `regex` already supports via no-default-features. This should be doable just by refactoring the feature list.

Currently, `lazy-regex`, with no-default-features, doesn't compile as it doesn't pull `regex` as it is an optional dependency.

**Comment by Canop** [comment]
This looks like two different problems.

1. Can you please prepare an example of no-std lib or program which doesn't compile with lazy-regex due to this std dependency ?

2. I'm not sure there's a way to have regex and regex-lite be mutually exclusive and allow compilation with `--no-default-features`. But how is that a problem ?

**Comment by rnbguy** [comment]
I didn't take a look at your source code before. Regarding point 2, it is possible to go around it.

```toml
[target.'cfg(all(not(feature = "default"), not(feature = "lite")))'.dependencies]
regex = {version = "1.9", default_features = false}
```

But it's not recommended since [`crate/feature`](https://doc.rust-lang.org/cargo/reference/features.html) has landed, so I agree with you. Also, Rust-Doc has a paragraph about [mutually exclusive features](https://doc.rust-lang.org/cargo/reference/features.html#mutually-exclusive-features).

About point 1, I don't have any easy/good examples[^1]. But the changes should be something like similar to this,

```diff
  [dependencies]
  ...
- regex = {version = "1.9", default_features = false, features = ["std"], optional = true}
+ regex = {version = "1.9", default_features = false, optional = true}

  [features]
  ...
+ no_std = ["regex"]
```

But again usually [`no_std` feature](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification) is also not recommended. But with mutually exclusive features - I don't think you have another choice.

[^1]: I was trying to use `lazy-regex` in our [`ibc` crate](https://crates.io/crates/ibc). I tried to use my fixes, but one of our dependencies locked on to an older version of `regex` than `lazy-regex` :man_facepalming:

**Comment by Canop** [comment]
To be honest the ` features = ["std"]` part in my regex import seems to be a forgotten relic and I think it has exactly no impact due to the rest but I may be missing some subtleties of dependencies computations.

**Comment by rnbguy** [comment]
> I think it has exactly no impact due to the rest

Exactly. That's why I turned it off - so that `no_std` doesn't pull `regex/std`. You already pull `regex/default` ([which includes `regex/std`](https://github.com/rust-lang/regex/blob/061ee815ef2c44101dba7b0b124600fcb03c1912/Cargo.toml#L33)) by your default feature.

Only some of the `regex/perf*` and `regex/unicode*` features may require `std`. So you will need to update them as follows,

```diff
  [features]
  ...
  std = ["regex/std"]
  ...
- perf = ["regex/perf"]
+ perf = ["std", "regex/perf"]
```

PS. I understand that [`regex-lite` requires `std` and doesn't support `no_std`](https://github.com/rust-lang/regex/blob/061ee815ef2c44101dba7b0b124600fcb03c1912/regex-lite/Cargo.toml#L19-L20). So no need to touch it for `no_std`.

**Comment by Canop** [comment]
Hum... I think this can be closed now until I miss something...
